caffe2/core/module.cc

```diff
@@ -0,0 +1,122 @@
+#include "caffe2/core/logging.h"
+#include "caffe2/core/module.h"
+
+#ifndef _MSC_VER
+#include <dlfcn.h>
+#endif
+
+namespace caffe2 {
+
+static std::mutex& gModuleChangeMutex() {
+  static std::mutex m_;
+  return m_;
+}
+
+static CaffeMap<string, const ModuleSchema*>& MutableCurrentModules() {
+  static CaffeMap<string, const ModuleSchema*> module_schema_map_;
+  return module_schema_map_;
+}
+
+// Note(jiayq): I am not sure whether the module handles are going to be used
+// as C2 uses modules via registration, but let's keep the handles at least.
+static CaffeMap<string, void*> CurrentModuleHandles() {
+  static CaffeMap<string, void*> module_handle_map_;
+  return module_handle_map_;
+}
+
+const CaffeMap<string, const ModuleSchema*>& CurrentModules() {
+  return MutableCurrentModules();
+}
+
+ModuleSchema::ModuleSchema(const char* name, const char* description)
+    : name_(name), description_(description) {
+  std::lock_guard<std::mutex> guard(gModuleChangeMutex());
+  MutableCurrentModules().emplace(name, this);
+}
+
+bool HasModule(const string& name) {
+ auto& modules = CurrentModules();
+ return (modules.find(name) != modules.end());
+}
+
+#ifdef _MSC_VER
+
+void LoadModule(const string& name, const string& filename) {
+  CAFFE_ENFORCE(!HasModule(name),
+    "On Windows, LoadModule is currently not supported yet and you should "
+    "use static linking for any module that you intend to use.");
+}
+
+#else
+
+void LoadModule(const string& name, const string& filename) {
+  CAFFE_ENFORCE(
+      name.size() > 0 || filename.size() > 0,
+      "You must provide at least one of name and filename.");
+  if (name.size() && HasModule(name)) {
+    VLOG(1) << "Module " << name << " already present. Skip loading."; 
+    return;
+  }
+  void* handle = nullptr;
+  if (filename.size()) {
+    handle = dlopen(
+        filename.c_str(), RTLD_NOW | RTLD_GLOBAL);
+    CAFFE_ENFORCE(handle != nullptr,
+      "Cannot load module ",
+      name,
+      " (with given filename ",
+      filename,
+      "), are you sure it is correct?");
+  } else {
+    string inferred_name = string("lib") + name + ".so";
+    handle = dlopen(
+        inferred_name.c_str(), RTLD_NOW | RTLD_GLOBAL);
```

Comment:
Is there a particular reason why this doesn't also use RTLD_NODELETE as its mode?

Reply:
Android does not seem to support it. I actually placed RTLD_NODELETE but then had to remove it.

In practice we would probably never dl close a module so this is a minor point, but if issues emerge we can macro define it in non-Android cases.